active/0051-strategies.md

```diff
@@ -0,0 +1,248 @@
++ Start Date: 2017-09-07
++ RFC PR: (leave this empty)
+
+# Summary
+
+The goal of this RFC is to solidify the next version of the mechanism that `ember-cli` uses to build final assets. It will allow for a more flexible build pipeline for Ember applications. It also unlocks building experimental features on top. It is a backward compatible change.
+
+# Motivation
+
+The [Packager RFC](https://github.com/chadhietala/rfcs/blob/packager/active/0002-packager.md) submitted by [Chad Hietala](https://github.com/chadhietala) is a little over 2 years old. A lot of things have changed since then and it requires a revision.
+
+The current application build process merges and concatenates input broccoli trees. This behaviour is not well documented and is a tribal knowledge. While the simplicity of this approach is nice, it doesn't allow for extension. We can refactor our build process and provide more flexibility when desired.
+
+Most importantly, it helps us achieve:
+
++ defining and developing a common language around the subject
++ removing highly coupled code and streamline technical implementation (Ember Engines)
++ unlock a whole different set of plugins we couldn't have before:
+  + ability to create custom bundles (i.e per-engine and per-route bundles)
+  + take advantage of [HTTP2 multiplexing](https://http2.github.io/faq/#why-is-http2-multiplexed) and [cache pushing](https://www.mnot.net/blog/2014/01/30/http2_expectations#4-cache-pushing)
+  + optimising plugins (Javascript and CSS treeshaking)
+
+
+
+# Scope
+
++ New public API for registering strategies
+
+# Terminology
+
++ **Strategy** - A strategy is responsible for returning a build pipeline that can emit a specific set of output assets given an input tree.
++ **Assembler** - Responsible for taking the resolved tree and applying – default or user provided – strategies to the tree.
+
+# Detailed design
+
+The detailed design is separated in various sections so that it is easier for a reader to understand.
+
+## Assembler
+
+`Assembler` is similar to a car assembly line - starting with a metal frame, making changes to it along the way, adding parts and at the end, there's an assembled car.
+
+It does not make any assumptions about how the final output should be constructed. Instead, it relies on strategies to tell it what the final output should look like. This is rather important as it enforces clear separation of concerns.
+
+Assembler must have the following interface:
+
+```typescript
+interface Assembler {
```

Comment:
Is this going to be a public API? So are there use cases for different `Assembler` implementations? Or is this just for clarification, but the end just an (internal) implementation detail?

Reply:
`Assembler` is more of a concept, than anything else. If you start thinking in terms of strategies, it is nothing more but another type of strategy. [As of right now](https://github.com/ember-cli/ember-cli/blob/master/lib/broccoli/assembler.js), it is an internal construct.